plugin/pulpcore/plugin/changeset/model.py

```diff
@@ -131,13 +135,42 @@ def save(self):
             except IntegrityError:
                 is_duplicate = True
             else:
-                for artifact in self.artifacts:
-                    artifact.model.content = self.model
-                    artifact.save()
-        if is_duplicate:
-            model = type(self.model)
-            content = model.objects.get(**self.key)
-            self.update(content)
+
+                for remote_artifact in self.artifacts:
+                    if not remote_artifact.content_artifact.artifact_id:
+                        if not deferred:
+                            artifact = self.validate_and_create_artifact(remote_artifact.model,
+                                                                         remote_artifact.download)
+                            remote_artifact.content_artifact.artifact = artifact
+                        remote_artifact.model.save()
+                        remote_artifact.content_artifact.save()
+
+    def validate_and_create_artifact(self, remote_artifact, download):
+        """
+        Create an Artifact from a RemoteArtifact and a downloaded file
+
+        Args:
+            remote_artifact (pulpcore.plugin.RemoteArtifact): remote artifact
+            download (download object): download
+            :return:
+        """
+        file = File(open(download.writer.path, mode='rb'))
+        file.hashers = download.writer.hashers
+
+        # Build data from RemoteArtifact
+        data = QueryDict('', mutable=True)
+        data['file'] = file
+        if remote_artifact.size:
+            data['size'] = remote_artifact.size
+        for hasher in hashlib.algorithms_guaranteed:
+            if getattr(remote_artifact, hasher, None):
+                data[hasher] = getattr(remote_artifact, hasher, None)
+
+        serializer_class = get_plugin_config("pulp_app").named_serializers["ArtifactSerializer"]
+        serializer = serializer_class(data=data)
+        serializer.validate(data=data)
+        serializer.is_valid(raise_exception=True)
+        return serializer.save()
```

Comment:
What about race conditions whereby the artifact does not exist when the artifact iterator checks but another thread has created it before this code tries to insert it.  The dup needs to be handled.